src/main/java/com/petros/bibernate/session/SessionImpl.java

```diff
@@ -1,58 +1,125 @@
 package com.petros.bibernate.session;
 
+import com.petros.bibernate.action.DeleteEntityAction;
+import com.petros.bibernate.action.EntityAction;
+import com.petros.bibernate.action.InsertEntityAction;
 import com.petros.bibernate.config.Configuration;
 import com.petros.bibernate.dao.EntityPersister;
 import com.petros.bibernate.exception.BibernateException;
 import com.petros.bibernate.session.context.PersistenceContext;
 import com.petros.bibernate.session.context.PersistenceContextImpl;
 
 import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Comparator;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
 
 public class SessionImpl implements Session {
     private boolean isOpened = true;
+    private final Transaction transaction;
+    private final DataSource dataSource;
     private final EntityPersister entityPersister;
-
+    private final Queue<EntityAction> actionQueue;
     private final PersistenceContext persistenceContext;
+    private Connection connection;
 
     public SessionImpl(DataSource dataSource, Configuration configuration) {
-        this.entityPersister = new EntityPersister(dataSource, configuration.showSql());
+        this.entityPersister = new EntityPersister(configuration.showSql());
+        this.dataSource = dataSource;
         this.persistenceContext = new PersistenceContextImpl();
+        this.actionQueue = new PriorityQueue<>(Comparator.comparing(EntityAction::priority));
+        this.transaction = new TransactionImpl(this);
     }
 
     @Override
     public void flush() throws BibernateException {
-        checkSession();
+        requireOpenSession();
+        openConnection();
+        setAutoCommitIfTxOpen(FALSE);
+        actionQueue.forEach(action -> action.execute(connection, persistenceContext));
+        actionQueue.clear();
+        setAutoCommitIfTxOpen(TRUE);
+    }
+
+    private void setAutoCommitIfTxOpen(boolean isAutoCommit) {
+        if (transaction.isOpen()) {
+            try {
+                connection.setAutoCommit(isAutoCommit);
+            } catch (SQLException ex) {
+                throw new BibernateException("Could not set autoCommit=false", ex);
+            }
+        }
     }
 
     @Override
-    public <T> void persist(T object) {
-        checkSession();
-        persistenceContext.cache(entityPersister.insert(object));
+    public <T> void persist(T entity) {
```

Comment:
Should we first check if the entity is already in the persistence context? If it is, throw an exception, and if not, perform all these actions along with adding the entity to the persistence context?

Reply:
I've implemented this behaviour during covering code with tests. Even more, we cannot remove entity, if it is not in persistent state (not in persistence context)